src/win/util.c

```diff
@@ -1299,20 +1242,34 @@ void uv_os_free_passwd(uv_passwd_t* pwd) {
 }
 
 
-int uv__convert_utf16_to_utf8(const WCHAR* utf16, char** utf8) {
+/*
+ * Converts a UTF-16 string into a UTF-8 one. The resulting string is
+ * null-terminated.
+ *
+ * If utf16 is null terminated, utf16len can be set to -1, otherwise it must
+ * be specified.
+ */
+int uv__convert_utf16_to_utf8(const WCHAR* utf16, int utf16len, char** utf8) {
   DWORD bufsize;
 
   if (utf16 == NULL)
     return UV_EINVAL;
 
   /* Check how much space we need */
-  bufsize = WideCharToMultiByte(CP_UTF8, 0, utf16, -1, NULL, 0, NULL, NULL);
+  bufsize = WideCharToMultiByte(CP_UTF8,
+                                0,
+                                utf16,
+                                utf16len,
+                                NULL,
+                                0,
+                                NULL,
+                                NULL);
 
   if (bufsize == 0)
     return uv_translate_sys_error(GetLastError());
 
   /* Allocate the destination buffer */
-  *utf8 = uv__malloc(bufsize);
+  *utf8 = uv__malloc(bufsize + 1);
```

Comment:
I don't think the `+ 1` is required in all cases. It's probably simpler and safer to just always allocate it though.